pkg/agent/agent.go

```diff
@@ -0,0 +1,111 @@
+package agent
+
+import (
+	"strconv"
+
+	"github.com/kubeapps/kubeapps/pkg/proxy"
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/kube"
+	"helm.sh/helm/v3/pkg/release"
+	"helm.sh/helm/v3/pkg/storage"
+	"helm.sh/helm/v3/pkg/storage/driver"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+)
+
+type DriverType string
+
+const (
+	Secret    DriverType = "SECRET"
+	ConfigMap DriverType = "CONFIGMAP"
+	Memory    DriverType = "MEMORY"
+)
+
+type Options struct {
+	ListLimit int
+	Timeout   int64
+}
+
+type Config struct {
+	ActionConfig *action.Configuration
+	AgentOptions Options
+}
+
+func ListReleases(config Config, namespace string, status string) ([]proxy.AppOverview, error) {
+	cmd := action.NewList(config.ActionConfig)
+	if namespace == "" {
+		cmd.AllNamespaces = true
+	}
+	cmd.Limit = config.AgentOptions.ListLimit
+	releases, err := cmd.Run()
+	if err != nil {
+		return nil, err
+	}
+	appOverviews := make([]proxy.AppOverview, len(releases))
+	for i, r := range releases {
+		appOverviews[i] = appOverviewFromRelease(r)
+	}
+	return appOverviews, nil
+}
+
+func NewActionConfig(driver DriverType, token, namespace string) *action.Configuration {
+	actionConfig := new(action.Configuration)
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		panic(err.Error())
+	}
+	config.BearerToken = token
+	config.BearerTokenFile = ""
+	clientset, err := kubernetes.NewForConfig(config)
+	store := createStorage(driver, namespace, clientset)
+	actionConfig.RESTClientGetter = nil     // TODO replace nil with meaningful value
+	actionConfig.KubeClient = kube.New(nil) // TODO replace nil with meaningful value
+	actionConfig.Releases = store
+	actionConfig.Log = klog.Infof
+	return actionConfig
+}
+
+func createStorage(driverType DriverType, namespace string, clientset *kubernetes.Clientset) *storage.Storage {
+	var store *storage.Storage
+	switch driverType {
+	case Secret:
+		d := driver.NewSecrets(clientset.CoreV1().Secrets(namespace))
+		d.Log = klog.Infof
+		store = storage.Init(d)
+	case ConfigMap:
+		d := driver.NewConfigMaps(clientset.CoreV1().ConfigMaps(namespace))
+		d.Log = klog.Infof
+		store = storage.Init(d)
+	case Memory:
+		d := driver.NewMemory()
+		store = storage.Init(d)
+	default:
+		// No (real) enums/ADTs in Go, so no static guarantee against this case.
+		panic("Invalid Helm driver type: " + driverType)
```

Comment:
In this specific case, I'm actually not sure about that. I wanted to have the first argument to `agent.createStorage` be of a sum type, so that I'd only need to handle exactly the cases represented by that sum type. There seems to be no sum types in Go, so said first argument instead has the quasi sum type `DriverType`. This at least means that values like `os.Getenv("LOL"))` don't typecheck:

```
cannot use os.Getenv("LOL") (type string) as type agent.DriverType in argument to handler.WithAgentConfig
```

However, values like `"hello"` apparently count as members of `DriverType`, but I found no way around that.

I think returning an error in `createStorage` would introduce a lot of (IMO) unnecessary complexity in `agent.NewActionConfig` and `handler.WithAgentConfig`. I also think a programmer will understand that they're not supposed to pass `"hello"`, but an actual `DriverType` (e.g. `agent.Secret`).

At some point, you want to be able to say "I have a value of type _T_ – I know it has been validated _because_ it is a _T_." [_Parse, don't validate._](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/) Go doesn't seem to help us more than I've already made it do.

I'm much more inclined to make `agent.ParseDriverType` return an error instead of panicking, because, unlike `createStorage`, it promises that it can handle any `string`. Its current deceitfulness stems from my interpretation of @absoludity's [suggestion](https://github.com/kubeapps/kubeapps/pull/1309#discussion_r349377228) that Kubeops should refuse to start if an invalid driver type has been specified. I'll be more than happy to return an error instead.

Reply:
If I understand this correctly, you mean that `driverType` was validated before so this won't never panic, right? (because the driver was parsed before). If that's the case then it's fine, I just found it weird for a function in a library to panic.

Reply:
Yeah, it is weird. But Go doesn't seem to let me express a good old sum type, so this was the best solution I could come up with. It is correct that `createStorage` should never panic, because `driverType` should really be a `DriverType`. The only way I know to violate this assumption is to pass a string _literal_ as `driverType`.